node/src/chain/acceptor.rs

```diff
@@ -65,31 +65,56 @@ impl<DB: database::DB, VM: vm::VMExecution, N: Network> Drop
 impl<DB: database::DB, VM: vm::VMExecution, N: Network> Acceptor<N, DB, VM> {
     pub async fn new_with_run(
         keys_path: &str,
-        mrb: &BlockWithLabel,
+        mrb: BlockWithLabel,
         provisioners_list: &Provisioners,
         db: Arc<RwLock<DB>>,
         network: Arc<RwLock<N>>,
         vm: Arc<RwLock<VM>>,
-    ) -> Self {
+    ) -> anyhow::Result<Self> {
+        let mrb_height = mrb.inner().header().height;
+        let mrb_state_hash = mrb.inner().header().state_hash;
+
         let acc = Self {
-            mrb: RwLock::new(mrb.clone()),
+            mrb: RwLock::new(mrb),
             provisioners_list: RwLock::new(provisioners_list.clone()),
             db: db.clone(),
             vm: vm.clone(),
             network: network.clone(),
             task: RwLock::new(Task::new_with_keys(keys_path.to_owned())),
         };
 
-        acc.task.write().await.spawn(
-            mrb.inner(),
-            provisioners_list,
-            &db,
-            &vm,
-            &network,
+        // NB. After restart, state_root returned by VM is always the last
+        // finalized one.
+        let state_root = vm.read().await.get_state_root()?;
+
+        info!(
+            event = "VM state loaded",
+            state_root = hex::encode(state_root),
         );
 
-        acc
+        // Detect a consistency issue between VM and Ledger states.
+        if mrb_height > 0 && mrb_state_hash != state_root {
+            info!("revert to last finalized state");
+            // Revert to last known finalized state.
+            acc.try_revert(RevertTarget::LastFinalizedState).await?;
```

Comment:
I can revert it, I thought this was a suggestion of your, but probably my memory serves bad

Reply:
> It's a bit misleading to call `try_revert` in `new_with_run` . Either rename the function name appropriately or at least put a description to document the additional call.

Please re-read. I suggested only to rename the function accordingly or document it so that it becomes obvious that we do more than initialize and run with calling it.

Reply:
Oh right, changed to `Acceptor::init_consensus` and added documentation